cpp/src/arrow/dataset/filter.cc

```diff
@@ -559,43 +558,35 @@ struct ScalarExpressionToString {
 };
 
 std::string ScalarExpression::ToString() const {
+  auto type_repr = value_->type->ToString();
```

Comment:
IDK that including this type is worthwhile (or if you think it is, then I probably want a different, simple print method that doesn't include it)

Reply:
Even with implicit casts, the precise type of a scalar in an expression is significant. Notably: without implicit casts `"b"_ > 4:int32` might be invalid where `"b"_ > 4:float64` is not. If you're looking at these strings trying to debug a casting error, you really need the type information for the scalars. IMHO, it's out of scope to provide a print function without this; it seems like you might be planning to use these for something other than debugging (like direct re-evaluation in R, maybe)?

Reply:
I wasn't planning on that, no. My intended use case was more "what is in this object anyway?" and not "why did my object not evaluate?".

Reply:
I would want the type to debug.

Reply:
🤷‍♂ I'll take it, it's a huge improvement.